### PR TITLE
Joining of indices

### PR DESCRIPTION
**Pull request by jijordre**
This PR corrects a problem seen when joining JSON-stat objects of class `dataset` with a defined value of `options.by`.

The signature of the experienced problem is that output index values coming from `arrobj` items at low index (of `arrobj`) are overwritten by `arrobj` items at higher index. It turns out that addition of index has to be treated differently than addition of label and unit.

In lack of existing unit test for the join module the fix was only tested by making use of the modified suite code in a third party solution. There the outcome when joining multiple JSON-stat data sets was successful. However, bear this lack of unit testing in mind when validating the PR.😄

### CODE REVIEW

**Comment by badosa** [comment]
@jijordre In lack of an existing unit test, could you please check how the change affects these examples

https://bl.ocks.org/badosa/5c788a54b4145d2b4c04dc718112af5d
https://bl.ocks.org/badosa/e1acf16f8a7ed871cff61d849706c202
https://bl.ocks.org/badosa/a1fb2b272a1f4a0bba9c041a4cf09078
https://bl.ocks.org/badosa/7245dccd2b8ea8a023c5c374a3fe0fe5

?

These examples worked correctly with jsonstat-utils@2 and I'm _assuming_ they work correctly also with jsonstat-suite@3.1.6.

**Comment by jijordre** [comment]
@badosa I tried out the first example you provided and found a strong lead on the root cause of the problem.

TL;DR The samples include to few categories in the dimension being joined and where `dimension.join_dimension.category.index` is non-array.

To explain I created this gist: [https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd). You may run the JavaScript file in NodeJS, after catering for its dependency to _axios_.

In the gist I query Statistics Norway's API. Files [`part0.json`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-part0-json) and [`part1.json`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-part1-json) are my two parts to be joined. The join dimension is `'Alder'`. The response in each part is limited to 3 categories for this dimension.

The thing to note is the non-array value of `dimension.Alder.category.index` if both [`part0.json`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-part0-json-L66) and [`part1.json`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-part1-json-L66). The values in each object overlap.

Then, using the `jsonstat-suite@3.1.6` to join I obtain the output in [`joined-3.1.6.json`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-joined-3-1-6-json). Here note the repeating values in [`dimension.Alder.category.index`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-joined-3-1-6-json-L66).

Shifting over to using the modified join of this PR the output is as obtained in [`joined-3.1.7.json`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-joined-3-1-7-json). There is no overlap in the values of [`dimension.Alder.category.index`](https://gist.github.com/jijordre/c204add7445aed4eb98be892cf515fdd#file-joined-3-1-7-json-L66).

So why did the samples you provided execute successfully? I noticed that your samples only have 1 category in the join dimension of each part when `dimension.join_dimension.category.index` is non-array, like in [https://bl.ocks.org/badosa/a1fb2b272a1f4a0bba9c041a4cf09078](https://bl.ocks.org/badosa/a1fb2b272a1f4a0bba9c041a4cf09078).

Hence I modified the gist JavaScript to only have 1 category in the join dimension of each part when `dimension.join_dimension.category.index` is non-array. This gives correct output also using `jsonstat-suite@3.1.6`. The reason for the correct output in this corner case (only 1 category in join dimension) seems to be the handing of `e[p]===0` in [`https://github.com/jsonstat/suite/blob/master/src/join.js#L54`](https://github.com/jsonstat/suite/blob/master/src/join.js#L54). That particular piece of logic makes the sample cases you provided go around.

Cheers

**Comment by badosa** [comment]
Thanks @jijordre for analyzing join() issues and providing a solution. The truth is join() never got a lot of interest (and those who were interested complained about its slowness). For me it was always something experimental, more conceptual than practical (that's why it uses an easy but inefficient approach): it was the way to put the focus on the topic of dataset arithmetics, a topic that the library should address if enough interest was aroused (which it didn't).

Can I ask you why and how are you using join() or how do you plan it?

**Comment by jijordre** [comment]
You're welcome, @badosa.

The use of join() in this context in related to queries of the API of Statistics Norway for a product called [Samla](https://www.webstep.no/vare-tjenester/samla/) ([Google-translated](https://translate.google.com/translate?sl=no&tl=en&u=https://www.webstep.no/vare-tjenester/samla/)).

The API has a constraint on the max amount of cells returned in any one request. Hence in order to query on large data sets we batch the categories in one dimension and to obtain partial data sets for each batch, which then subsequently is joined with jsonstat-suite.
